### ISSUE TITLE
[ROCm] gfx950 top-k policy is gated on topK==1024, leaving index_topk=2048 models (GLM-5.x) on the generic 10-way split

### ISSUE THREAD

**Issue by stefanskiasan** [vllm-project/vllm#55327]
## Summary

The measured gfx950 policy for the sparse-attention indexer's top-k, added in #52882, is gated on `topK == 1024`. Models with `index_topk: 2048` — GLM-5.2, GLM-5.3 — never reach it and fall back to the generic `multipleBlocksPerRowConfig = 10` with 512 threads per split block.

On MI355X that is the wrong choice for **every shape we measured**. A policy calibrated for `topK == 2048` gives a median 25 % and up to 64 % reduction in kernel time, with no shape getting slower and identical results.

## Where

`csrc/libtorch_stable/sampler.cu`, `top_k_per_row_decode`:

```cpp
const bool useGfx950DeviceLengthAwareTopK1024 =
    topK == 1024 && numRows > 64 && numRows <= 256 && gfx950;   // ~line 754
const bool useGfx950TopK1024 =
    topK == 1024 && numRows <= 64 && gfx950;                    // ~line 793
int multipleBlocksPerRowConfig = 10;                            // everything else
```

## Why the gap is larger, not smaller, at topK=2048

The merge kernel walks `B * topK` candidates in a **single** block per row:

```cpp
vllm::topKPerRowDecode<1024, true, false, true>
    <<<numRows, 1024, ...>>>( ..., multipleBlocksPerRowConfig * topK, ... );
```

At topK=1024 with B=10 that is 10 240 candidates; at topK=2048 it is **20 480**. The cost #52882 set out to remove is twice as large for the models that do not get the fix.

rocprofv3, 4 rows / 2 000 active columns, MI355X, GLM-5.3 (`index_topk: 2048`):

| kernel | grid | median |
|---|---|---|
| `topKPerRowDecode<1024,true,false,true>` (merge) | 4 × **1** | **95.4 µs** |
| `topKPerRowDecode<512,true,true,false>` (split) | 4 × 10 | 5.0 µs |

The split is irrelevant; the merge is the entire cost. In a full decode profile this kernel held 12 % of GPU time.

## Short contexts are the worst case, not long ones

`numColumns` is `logits.size(1)`, and the indexer's logits are allocated at `max_model_len` (`fp8_fp4_paged_mqa_logits(..., max_model_len=max_model_len)`). With a 1M model length the two-stage path is therefore taken **always**, including for a 2 000-token request — where the merge sorts 20 480 candidates to find the best 2 000 that exist.

Measured, same 4 rows: **2 000 tokens cost 112 µs, 116 000 tokens cost 63 µs.** The short context is nearly twice as expensive as the long one.

## Measurements

Micro-benchmark calling `torch.ops._C.top_k_per_row_decode` directly, checked against `torch.topk` on every point. 35 shapes (4–256 rows × 2k–600k active columns) × 9 split counts × 3 runs, plus a run with 1024 threads. MI355X, ROCm 7.2, AITER 0.1.21.

Best split count per shape (512 threads):

| rows | 2 000 | 8 000 | 32 000 | 64 000 | 116 000 | 256 000 | 600 000 |
|---|---|---|---|---|---|---|---|
| 4 | 2 | 4 | 5 | 5 | 6 | 10 | 16 |
| 32 | 2 | 4 | 16 | 6 | 8 | 8 | 12 |
| 64 | 2 | 4 | 3 | 4 | 4 | 8 | 8 |
| 128 | 2 | 4 | 2 | 2 | 4 | 4 | 6 |
| 256 | 2 | 4 | 2 | 2 | 3 | 3 | 4 |

Two effects, both already exploited by #52882 for topK=1024:

1. **Split count.** Short rows want few blocks (the merge dominates); long rows with few of them want many (the split dominates); many rows want few (the device is already saturated).
2. **Threads per split block.** 1024 instead of 512 is worth a further **+8 %** on average — and it shifts the optimum toward *smaller* split counts, so the two must be tuned together. Our first attempt reused the split counts measured at 512 threads and was consequently *slower* than baseline at long contexts.

**Separate finding: `B = 1` is not merely slow, it is incorrect** — 137 of 2048 indices match `torch.topk`. Every other value we measured (2…16) is exact. This is reachable today by anyone editing `multipleBlocksPerRowConfig`, and may deserve a guard regardless of this issue.

## Policy we are running

Same structure as `topKPerRowDecodeDeviceLengthAware`: grid and workspace sized from `numRows` only (capture-stable), active block count decided **on device** from the real row length.

```
rowEnd <  4 096                     -> 2 blocks
rowEnd < 16 384                     -> 4 blocks
otherwise: numRows <= 32 -> 6, <= 64 -> 4, else 2
split launched with 1024 threads
```

Against the current generic path, same 35 shapes:

* mean **+26.2 %**, median **+24.6 %**, range **+0.4 % … +63.7 %**
* **no shape slower**
* identical results everywhere (2048/2048 matching indices, max |Δvalue| = 0)
* within 1.7 % of the per-shape optimum on average

End-to-end on GLM-5.3 (TP4, MI355X, same build, only the policy toggled): single user +4.8 % tok/s, 64 concurrent +3.2 % and +3.9 % (two seeds, paired), GSM8K 96.8 % → 97.0 % over 1319 problems, needle-in-haystack unchanged.

## Questions

1. Was the `topK == 1024` gate deliberate — is there a correctness reason we are missing — or simply the scope of what was measured for DeepSeek-V4?
2. Would you prefer a second calibrated branch for 2048, or a table indexed by `(topK, numRows, rowEnd)` covering both?
3. Should `B = 1` be rejected outright given it returns wrong results?

Happy to open a PR with the policy above plus the micro-benchmark, if that is useful.

---

Related, separate observation: `sparse_attn_indexer.py` gates `use_persistent_topk` / `use_cooperative_topk` on `current_platform.is_cuda()`, so on ROCm every `index_topk` in (512, 1024, 2048) falls through to `top_k_per_row_decode`. That is expected — the CUDA kernels do not port — but it means the generic path carries all ROCm DSA traffic.
